Adds support for custom planets and General updates

Thutmose opened the pull request:
This pull request should address #184, #165 and #179.

For #184 and #165, which appear to be duplicates:

This adds a planets.json in `config/beyond_earth/`
This planets.json allows defining various dimension ids as planet/orbit pairs, and these can be arranged in solar systems and moon systems.
So far these planets have the various things that the default planets had, such as differing gravity.
Custom planets must still be manually added via datapack/etc, along with an orbit dimension for them.
Custom planet icons/etc must be included in a texture pack.

For #179 this adds:

`Planets.getLocationForOrbit` and `Planets.getLocationForPlanet` for quickly finding mappings of planet to orbit.

# Important Commits after the 10th:

- 11 Updates recipe types to 1.19.2
- 13 Updates AbstractMachineBlockEntity
- 14 Updates Coal Generator
- 16 Updates Compressor
- 17 Adds initial coal generator JEI stuff

## Notes:

Block/item models for machines were not updated here, just the functionalitiy.
JEI for coal generator hasn't had tooltips added yet.

Thutmose committed:
Update Jei.java

template for JEI module once recipes are in

Thutmose committed:
Step 1 of cleaning up level registry to more easily support adding planets

This moves most of the server-side references to Planet objects, which can be further expanded to load in from data, etc.

Thutmose committed:
More work on setting up a data-structure for loading systems

This will be able to do the following:

Load in a set of star systems, each containing planets, moons, etc.

Thutmose committed:
rework of more stuff related to planet registry

This does the following:

Star/planet buttons in the guis are now dynamically generated
IDs for the planets for buttons are synced to client with datapacks
Planets are arranged as listed in the planets file
Auto-generates a default planets file with the standard layout

Thutmose committed:
removes some debug prints

Thutmose committed:
Update Methods.java

changes the exiting of orbit to apply above min build height, instead of 0

Thutmose committed:
moves the rain, oxygen and "spacelevel" to the planets as well

gisellevonbingen commented:
How can I add my custom planet?
Looks require post some event on 'Planet#generateDefaults' called

How about change to load PlanetData from each json file in datapack?
Like recipe json files.
For don't create be file conflict and modpack/addon dev add star easier.

Thutmose commented:
The addon based system should still work as it did before, as the `PlanetSelectionScreenInitEvent` events are still called as they used to be. This system is for making the "default" set of planets customisable.

You can add the custom planets for a modpack by adding an appropriate entry to the `planets.json` that this generates in `config/beyond_earth/`

Here is the default json at present:
```json
{
  "stars": [
    {
      "name": "sun",
      "texture": "beyond_earth:textures/environment/planet/sun.png",
      "planets": [
        {
          "name": "mercury",
          "planet": "beyond_earth:mercury",
          "orbit": "beyond_earth:mercury_orbit",
          "planet_bar": "beyond_earth:textures/planet_bar/mercury_planet_bar.png",
          "texture": "beyond_earth:textures/environment/planet/mercury.png",
          "planet_item_gravity": 0.05,
          "planet_enitity_gravity": 0.02,
          "orbit_item_gravity": 0.05,
          "orbit_enitity_gravity": 0.01,
          "orbit_radius": 0.39,
          "mass": 0.055,
          "g": 1.0,
          "has_oxygen": false,
          "has_rain": true,
          "space_level": true,
          "rotation": 270.0,
          "orbit_colour": [
            179,
            49,
            44
          ],
          "tier": 3,
          "moons": []
        },
        {
          "name": "venus",
          "planet": "beyond_earth:venus",
          "orbit": "beyond_earth:venus_orbit",
          "planet_bar": "beyond_earth:textures/planet_bar/venus_planet_bar.png",
          "texture": "beyond_earth:textures/environment/planet/venus.png",
          "planet_item_gravity": -1.0,
          "planet_enitity_gravity": -1.0,
          "orbit_item_gravity": 0.05,
          "orbit_enitity_gravity": 0.01,
          "orbit_radius": 0.72,
          "mass": 0.81,
          "g": 1.0,
          "has_oxygen": false,
          "has_rain": true,
          "space_level": true,
          "rotation": 180.0,
          "orbit_colour": [
            235,
            136,
            68
          ],
          "tier": 3,
          "moons": []
        },
        {
          "name": "earth",
          "planet": "minecraft:overworld",
          "orbit": "beyond_earth:earth_orbit",
          "texture": "beyond_earth:textures/environment/planet/earth.png",
          "planet_item_gravity": -1.0,
          "planet_enitity_gravity": -1.0,
          "orbit_item_gravity": 0.05,
          "orbit_enitity_gravity": 0.01,
          "orbit_radius": 1.0,
          "mass": 1.0,
          "g": 1.0,
          "has_oxygen": true,
          "has_rain": true,
          "space_level": false,
          "rotation": 90.0,
          "orbit_colour": [
            53,
            163,
            79
          ],
          "tier": 1,
          "moons": [
            {
              "name": "moon",
              "planet": "beyond_earth:moon",
              "orbit": "beyond_earth:moon_orbit",
              "planet_bar": "beyond_earth:textures/planet_bar/moon_planet_bar.png",
              "planet_item_gravity": 0.05,
              "planet_enitity_gravity": 0.02,
              "orbit_item_gravity": 0.05,
              "orbit_enitity_gravity": 0.01,
              "orbit_radius": 1.0,
              "mass": 1.0,
              "g": 1.0,
              "has_oxygen": false,
              "has_rain": true,
              "space_level": true,
              "rotation": 0.0,
              "orbit_colour": [
                255,
                255,
                255
              ],
              "tier": 0,
              "moons": []
            }
          ]
        },
        {
          "name": "mars",
          "planet": "beyond_earth:mars",
          "orbit": "beyond_earth:mars_orbit",
          "planet_bar": "beyond_earth:textures/planet_bar/mars_planet_bar.png",
          "texture": "beyond_earth:textures/environment/planet/mars.png",
          "planet_item_gravity": 0.06,
          "planet_enitity_gravity": 0.05,
          "orbit_item_gravity": 0.05,
          "orbit_enitity_gravity": 0.01,
          "orbit_radius": 1.52,
          "mass": 0.107,
          "g": 1.0,
          "has_oxygen": false,
          "has_rain": true,
          "space_level": true,
          "rotation": 0.0,
          "orbit_colour": [
            37,
            49,
            146
          ],
          "tier": 2,
          "moons": []
        }
      ],
      "location": [
        0.0,
        0.0,
        0.0
      ],
      "colour": [
        255,
        255,
        255
      ],
      "mass": 1.0
    },
    {
      "name": "proxima_centauri",
      "planets": [
        {
          "name": "glacio",
          "planet": "beyond_earth:glacio",
          "orbit": "beyond_earth:glacio_orbit",
          "planet_bar": "beyond_earth:textures/planet_bar/glacio_planet_bar.png",
          "texture": "beyond_earth:textures/environment/planet/glacio.png",
          "planet_item_gravity": 0.05,
          "planet_enitity_gravity": 0.03,
          "orbit_item_gravity": 0.05,
          "orbit_enitity_gravity": 0.01,
          "orbit_radius": 0.39,
          "mass": 0.08,
          "g": 1.0,
          "has_oxygen": false,
          "has_rain": true,
          "space_level": true,
          "rotation": 180.0,
          "orbit_colour": [
            37,
            49,
            146
          ],
          "tier": 4,
          "moons": []
        }
      ],
      "location": [
        4.25,
        0.0,
        0.0
      ],
      "colour": [
        255,
        255,
        255
      ],
      "mass": 0.122
    }
  ]
}
```

You can add new planets via config by appending the appropriate planet object inside one of the "planets" lists, or add a new star by adding the appropriate star object inside the stars list. Adding moons consists of adding a planet to the moons list of another planet.

Adding a new default plant via pull request would be done in `Planet#generateDefaults`. I can see about adding an event which is fired at the end of that so addons can modify the default set, It would also need to be fired in `PlanetData#loadOrGenerateDefaults` if you want it to be added to the loaded planets as well.

Thutmose committed:
fixes orbit lines colours

Thutmose committed:
Update PlanetSelectionScreen.java

adds missing conversion from g to m/s^2

Thutmose committed:
adds events for when planets are generated/loaded

This should allow addons to modifiy the default set of planets

Thutmose commented on src/main/java/net/mrscauthd/beyond_earth/common/events/Events.java:
```diff
@@ -236,4 +233,106 @@ public static void livingSprinting(LivingSprintingEvent event) {
+     * 
+     * @param event
+     */
+    public static void onPlanetsGenerate(PlanetRegisterEvent.Generate event) {
```
@gisellevonbingen I added 2 events, `PlanetRegisterEvent#Generate` and `PlanetRegisterEvent#Load`, the base mod will use the former to add planets, as done here, addons can either use this event, or `PlanetRegisterEvent#Load` to also add their planets, the way it is done here should work as an example for both cases

gisellevonbingen replied:
PlanetSelectionScreenInitEvent is looks coarse
Branching on a 'category ID' integer variable is very outdated
Minecraft also changed block/item ID from integer to ResourceLocation

Best in my think is 'beyond earth'  be managing planet's relationship.
If register start system, planet and moons data in config.json or other ways,
It have to be show automatically in PlanetSelectionGUI

just my thoughts
If you think you don't have to, that's fine.

Thutmose replied:
I didn't make any changes to that, since that was part of the older API code.

My changes were to the internal, non-addon facing parts of the planet system. I did remove the giant wall of switch-case statements for the IDs though.

In the case for internal use however, the "category ID" is fine though, as that is a sequential system anyway, as it loops through the stars -> planets in the correct order

gisellevonbingen replied:
OK, i had look code some more,
Most of them seem to be
https://github.com/MrScautHD/Beyond-Earth/pull/185/files#diff-53634af117cdb210e31fb45406626b30d5a4a14bbb459c5114e8a50571456be3R240-R260

Thutmose replied:
that link seems to be corrupted, it just takes me to the top of the diffs, which file?

gisellevonbingen replied:
Ah never mind
i was wrong see it
https://github.com/MrScautHD/Beyond-Earth/blob/003ed4b980d2dc200fdc61057b88a9fc12be8fa9/src/main/java/net/mrscauthd/beyond_earth/client/screens/planetselection/PlanetSelectionScreen.java#L242-L260

Thutmose replied:
ya, they start at 0, then increase as you go to planets, it doesn't increase for moons/orbits.

This allows a reasonable way to maintaining the tree for finding which button should do what. It is also not user-facing, so numeric IDs are fine, and the numeric IDs can be assigned easily and automatically

Thutmose committed:
updates the recipe types and serialisers

Thutmose committed:
moves the recipes inside common

Thutmose committed:
Step 1 of machine updating

This at least updates the parts required for AbstractMachineBlockEntity.

Thutmose committed:
Updates coal generator

Thutmose committed:
adjustments to coal generator energy bar

Thutmose committed:
Updates compressor

Thutmose committed:
coal generator in JEI

Thutmose committed:
Update PlanetData.java

fixes g not synchronising from config

Thutmose committed:
Update ConfigSteelManagementCondition.java

moves this to common

Thutmose committed:
Update CoalGeneratorScreen.java

adjusts the location

Thutmose committed:
re-adds block texture/model for compressor and coal gen

MrScautHD commented:
amazing work!

MrScautHD commented:
i wrote you on dc what should improved. 👍

Thutmose committed:
removing these for using the new ones

Thutmose committed:
JEI for compressor recipes

still needs tooltips and possibly extra text additions for time, etc

Thutmose committed:
functional fuel refinery

Thutmose committed:
Oxygen Loader Functionality

Thutmose committed:
First step of solar panel update (some gui text is broken)

Thutmose committed:
nasa workbench updated

Thutmose committed:
O2 bubble distributor initial work

Thutmose committed:
Water pump re-added

Thutmose committed:
cleans up a variety of compiler warnings

Thutmose committed:
more on the oxygen bubble distributor

Thutmose committed:
some more compiler warning cleanup

Thutmose committed:
fixes size of fluid displays. Also adds initial JEI for the refinery

Thutmose committed:
More JEI related stuff

this should have all except for the nasa workbench mostly functional.

The textures still need updating to different ones if needed, for now they are just using the single texture.

This supports the fluid tanks draining/filling, the energy tanks draining/filling, using the same size as the standards in the inventory screens

Thutmose committed:
Update NASAWorkbench.java

JEI layout for this, note that the tags are not set right, so these recipes show barrier blocks as invalid for now

MrScautHD commented:
pls remove the stuff that already in the mod just with other names

MrScautHD commented:
like ImageButtonPlacer... it is ModifiedButton now (check the code before you add it)

Thutmose commented:
Do you have a list of what got renamed? I did look in generally similar named locations, but I guess I missed some of them

Thutmose commented:
It also appears that your "ModifiedButton " is much more limited than the older one, should I just adjust "ModifiedButton" to account for the extra textures, or should I just make a different button class for it?

Thutmose committed:
adjustments to helper button classes

Makes a more general "TexturedButton" for uses in things like the O2 bubble machine. The ModifiedButton then extends that, and is for use in the planet selection screens.

Thutmose committed:
adjustments to configurable planet stuff

this removes some of the extra gravity settings, these can all by handled by the "g" value for the planet

Thutmose committed:
more adjustments to simplify the config

there is now a g value for orbit, to allow changing it to something other than the normal amount.

Thutmose committed:
initial framework for artificial gravity blocks

this allows us to make an artificial gravity block, which can be used to make it nice to walk around on stations/moons/etc, or to work as interesting traps.

Thutmose committed:
changes space suit stuff to using a tag

this allows other mod's stuff to be set as space suits, without them having to have their armour actually depend on this mod.

Example: Pokecube adds ultrasapce suits, which can now just be tagged accordingly.

Thutmose committed:
fall damage based on g as well now

this removes yet another set of random numbers needed for each planet

Thutmose committed:
Basic implementation of room-based O2 distributor

This presently works as follows:

O2 distributor block produces O2, sends out of the block.
O2 will distribute around rooms when produced by the blocks, and will stop at solid walls.
O2 will also somewhat be produced by leaf blocks, but nowhere near as fast as O2 distributor.

This system is designed to be able to be extended further later, to properly handle doors, etc.

Thutmose committed:
rename to OxygenDistributor

Also makes it so torches/etc will not burn out if placed somewhere with O2 from the distributor

Thutmose committed:
adds rendering for moons around planets in the planet selection screen

Also some disabled stuff for phobos and deimos which was used for testing

Thutmose committed:
Standardises O2 breathing

This simplifies the needed code for making sure that entities/players can or cannot breathe at a location.

Thutmose committed:
more adjustments for oxygen distributor

rooms filled by the distributor should now also prevent burning on mercury/vanus

Thutmose committed:
Update OxygenSystem.java

removes a debug print

Thutmose committed:
re-organise these

Moves most of this ported inventory stuff into sub-folders

Thutmose committed:
removes un-needed O2 effect

Thutmose committed:
Update TexturedButton.java

cleanup of this

Thutmose committed:
Re-work of planet sky rendering

This does the following:

Renders planets and moons in the sky.
If it is rendering a moon, and is set as tidally locked, it renders the parent body in a fixed location based on the x-coordinate.
Moons of the current planet will render, and if they have a phase texture, it will use that, with a phase based on relative angle of the sun and the moon.
Planets will move throughout the sky depending on their location in orbit around the star.
Other moons will also appear in the sky if you are on a Moon

Configurable planet file can be used to adjust things like:

Whether clouds render on the planet.
Whether rain renders on the planet.
Offsets for sunrise/sunset colours

Thutmose committed:
makes planets in the sky fainter during the day

this makes it to closer to noon, the fainter the planets get

Thutmose committed:
Update SkyHelper.java

support for different light colours for configured stars

Thutmose committed:
Update Planets.java

Thutmose committed:
Makes some of the suggested re-organisations

Moves planet config stuff to data
Moves crafting to data, renames as recipes
Moves power system and gauge things to blocks/entities/machines
Moves GridPlacer to nasaworkbench

MrScautHD commented:
did you the changes already? (that we talked on dc)

Thutmose commented:
> 

I made several of the re-organisation changes, such as moving some of the helper classes around, that was done on the most recent commit: https://github.com/MrScautHD/Beyond-Earth/pull/185/commits/93c064a2bb2f2d5a7e37fb0505e721a62f378e51

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MrScautHD/Beyond-Earth?pullRequest=185) <br/>All committers have signed the CLA.

MrScautHD commented:
thx! amazing work!
